source/extensions/tracers/opentelemetry/http_trace_exporter.cc

```diff
@@ -0,0 +1,80 @@
+#include "source/extensions/tracers/opentelemetry/http_trace_exporter.h"
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "source/common/common/enum_to_int.h"
+#include "source/common/common/logger.h"
+#include "source/common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenTelemetry {
+
+OpenTelemetryHttpTraceExporter::OpenTelemetryHttpTraceExporter(
+    Upstream::ClusterManager& cluster_manager,
+    const envoy::config::core::v3::HttpService& http_service)
+    : cluster_manager_(cluster_manager), http_service_(http_service) {}
+
+bool OpenTelemetryHttpTraceExporter::log(const ExportTraceServiceRequest& request) {
+  std::string request_body;
+
+  const auto ok = request.SerializeToString(&request_body);
+  if (!ok) {
+    ENVOY_LOG(warn, "Error while serializing the binary proto ExportTraceServiceRequest.");
+    return false;
+  }
+
+  const auto thread_local_cluster =
+      cluster_manager_.getThreadLocalCluster(http_service_.http_uri().cluster());
+  if (thread_local_cluster == nullptr) {
+    ENVOY_LOG(error, "OTLP HTTP exporter failed: [cluster = {}] is not configured",
+              http_service_.http_uri().cluster());
+    return false;
+  }
+
+  Http::RequestMessagePtr message = Http::Utility::prepareHeaders(http_service_.http_uri());
+
+  // The request follows the OTLP HTTP specification:
+  // https://github.com/open-telemetry/opentelemetry-proto/blob/v1.0.0/docs/specification.md#otlphttp.
+  message->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
+  message->headers().setReferenceContentType(Http::Headers::get().ContentTypeValues.Protobuf);
+
+  // Add all custom headers to the request.
+  for (const auto& header_value_option : http_service_.request_headers_to_add()) {
+    message->headers().setCopy(Http::LowerCaseString(header_value_option.header().key()),
+                               header_value_option.header().value());
+  }
+  message->body().add(request_body);
+
+  const auto options = Http::AsyncClient::RequestOptions().setTimeout(std::chrono::milliseconds(
+      DurationUtil::durationToMilliseconds(http_service_.http_uri().timeout())));
+
+  Http::AsyncClient::Request* http_request =
+      thread_local_cluster->httpAsyncClient().send(std::move(message), *this, options);
+
+  return http_request;
```

Comment:
Got it. But please don't do that. If a bool return value is necessary, please use `return http_request != nullptr;`

But I think there are another more important problem. Because this http request is asynchronous, it's possible the exporter is destroyed before the request is completed. Then the request callback `*this` (Http::AsyncClient::Callbacks) will be a dangling reference.
If you do concern this exporting result, then you must manage the `http_request` and when the exporter is destroyed, you should cancel all pending request. 
If you don't care the sending result, than please use a static noop callback here to replace `*this` to avoid possible dangling reference.

Reply:
Got it. But please don't do that. If a bool return value is necessary, please use return http_request != nullptr;

But I think there is another more important problem. Because this http request is asynchronous, it's possible the exporter is destroyed before the request is completed. Then the request callback *this (Http::AsyncClient::Callbacks) will be a dangling reference.
If you do concern this exporting result, then you must manage the http_request and when the exporter is destroyed, you should cancel all pending request.
If you don't care the sending result, than please use a static noop callback here to replace *this to avoid possible dangling reference.

Reply:
Thanks for spotting this! 

I feel it's relevant to know the result. I looked at the Zipkin tracer and found they keep track of the "in flight" requests via `Http::AsyncClientRequestTracker`. I looked into `async_client_utility.cc` and that seems to properly cancel the requests. Also since the tracer/exporter are inside the TLS, no synchronization is needed on Http::AsyncClientRequestTracker. 

I also fixed the implict boolean cast, I think it's much better now! Thanks [e339153](https://github.com/envoyproxy/envoy/pull/29207/commits/e3391535f9c19ff7f4cf941b4c1ab4bf6a61ee3a)